### ISSUE TITLE
Windows install fails with "rmdir: <...> Permission denied"

### ISSUE THREAD

**Issue by db4** [ocaml/dune#2324]
OPAM dune install under Windows now fails with the following log:

```
Error: rmdir: _boot/.sandbox/429c3326ca2b6cc28ac5705de6baa034\build\default: Permission denied
-> required by src/dag/.dag.objs/native/dag__.cmx
-> required by src/.dune.objs/native/dune__Action.cmx
-> required by bin/.main.objs/native/main.cmx
-> required by bin/main.cmxa
-> required by bin/main/main_dune.exe
-> required by install bin/dune.exe
-> required by dune.install
Error: rmdir: _boot/.sandbox/c2c3a2f577a78d4b6194614ec945cb28\build\default: Permission denied
-> required by src/.dune.objs/native/dune.cmx
-> required by bin/.main.objs/native/main.cmx
-> required by bin/main.cmxa
-> required by bin/main/main_dune.exe
-> required by install bin/dune.exe
-> required by dune.install
Error: rmdir: _boot/.sandbox/7516a95ca4df5bdf8ce3236877b9a905\build\default: Permission denied
-> required by src/dune_lang/.dune_lang.objs/native/dune_lang__.cmx
-> required by src/.dune.objs/native/dune__Action.cmx
-> required by bin/.main.objs/native/main.cmx
-> required by bin/main.cmxa
-> required by bin/main/main_dune.exe
-> required by install bin/dune.exe
-> required by dune.install
```

Dune version does not matter: I tried 1.10.0, 1.9.0, 1.8.0, 1.7.3 with the same result. Interestingly, it worked for me in the past for the same OPAM install (although failed sometimes), but now after unsuccessful attempt to upgrade to version 1.10.0 it fails all the time. Looks like to the moment of `rmdir` some files may remain open in `default` directory. For Unix it does not matter but under Windows "permission denied" error is raised.

Any idea how to find and fix this error? It's real show-stopper for me.

**Comment by db4** [comment]
Another note: `./boot.exe --release -j 3` fails with the error above but `./boot.exe --release -j 1` succeeds.

**Comment by rgrinberg** [comment]
No idea why you're seeing this failure as it can't be a permission issue. The files and directories in .sandbox are created by dune itself.

Paths like this look a little suspicious to me:

```
Error: rmdir: _boot/.sandbox/7516a95ca4df5bdf8ce3236877b9a905\build\default
```

Why are mixing different slashes? @dra27 is this ok?

**Comment by nojb** [comment]
No idea about the particular problem, but on Windows, you can't delete a file that is is use. So race conditions can cause such issues (which explains why `-j1` makes the problem disappear).

**Comment by dra27** [comment]
"Permission denied" is the obnoxious error message for "file/directory in use". The shell at least now displays "The process cannot access the file because it is being used by another process.".

The other reason this can happen is if there are still files in the directory (again, the shell says "The directory is not empty.", but I've got a feeling the generic error when doing that from a process is permission denied). This can also happen, especially when `.exe` files are involved, if a virus scanner / anti-malware scanner is still scanning the file. That process is exceptionally weird, since the file is deleted, but hasn't yet disappeared because a system process is still working on it, but it will disappear as soon as the scan completes (we used to have to work around this in the OCaml testsuite where multiple programs were compiled successively with the same `program.exe` filename). That might be another explanation for why `-j 3` exposes the problem - the system is under more load.
